### PR TITLE
Domino refactor

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

## Description
This PR refactors the DoMINO model code for a few reasons:
- First, the model.py file was quite lengthy, and hard to follow.  By breaking it down into separate files, it's marginally better, though that's not the primary reason for the PR.
- The MLP implementation is predominantly consolidated, excluding `GeoConvOut`, which follows a different pattern.  This makes it easier to optimize and manage all the MLP implementations in one place.
- The solution calculation functions were refactored to ensure that only torch.Tensor inputs appear in the inputs of the functions.  This will make it easier to compile and optimize those functions - and on Hopper and newer, those are the main bottlenecks.
- Finally, a more extensive test suite validating components individually has been added.  The tests only check that the pieces can be constructed and produce expected output shapes, they don't check numerical stability.  But, it does have a more comprehensive coverage of the model parameters, including things that weren't previously tested (and failed, initially).

Some minor tweaks to the model have been made in a few small places:
- The `volume_min_max` and `surface_min_max` pieces of the input dict now are only used if they are present in the input dict.  This is with an eye to the dataloader refactor, which will not put those in the input dict if normalization is not in use.  That leads to consistent normalization.
- The solution calculation for surface data was breaking if `use_surface_normals=False` and `use_surface_area=True`.  The the area factors were enclosed within a `use_surface_normals` check, which is fixed.

## Checklist

- [X] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.
- [X] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Dependencies

No new dependencies.

### CODE REVIEW

File: physicsnemo/models/domino/geometry_rep.py [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+import torch.nn.functional as F
+from einops import rearrange
+
+from physicsnemo.models.unet import UNet
+
+from .ball_query import BQWarp
+from .encodings import fourier_encode_vectorized
+
+
+def get_activation(activation: Literal["relu", "gelu"]) -> Callable:
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/geometry_rep.py]: I think get_activation() needs to be defined in model.py

File: physicsnemo/models/domino/encodings.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    freqs = freqs[None, None, :, None]  # reshape to [*, F, 1] for broadcasting
+
+    coords = coords.unsqueeze(-2)  # [*, 1, D]
+    scaled = (coords * freqs).reshape(*coords.shape[:-2], D * F)  # [*, D, F]
+    features = torch.cat([torch.sin(scaled), torch.cos(scaled)], dim=-1)  # [*, D, 2F]
+
+    return features.reshape(*coords.shape[:-2], D * 2 * F)  # [*, D * 2F]
+
+
+class EncodingMLP(nn.Module):
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/encodings.py]: Should this be part of mlps.py because it is a different variant of an MLP with fourier encodings? We can call it something like Fourier MLP or keep the name.

File: physicsnemo/models/domino/model.py [hunk 3]
```diff
@@ -1912,15 +543,20 @@ def forward(self, data_dict, return_volume_neighbors=False):
...
+            if "surface_min_max" in data_dict.keys():
+                surf_max = data_dict["surface_min_max"][:, 1]
+                surf_min = data_dict["surface_min_max"][:, 0]
+                geo_centers_surf = (
+                    2.0 * (geo_centers - surf_min) / (surf_max - surf_min) - 1
+                )
+            else:
+                geo_centers_surf = geo_centers
+
             encoding_g_surf = self.geo_rep_surface(
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/model.py:555]: I saw a geo_rep_surface1 defined in the constructor of Domino. I think it is not used anywhere. We can remove it.

File: physicsnemo/models/domino/model.py [hunk 4]
```diff
@@ -1210,81 +378,33 @@ def __init__(
...
-                LocalPointConv(
-                    input_features=total_neighbors_in_radius,
-                    base_layer=512,
-                    output_features=self.volume_neighbors_in_radius[ct],
-                    model_parameters=model_parameters.local_point_conv,
-                )
-            )
 
         # Transmitting surface to volume
         self.surf_to_vol_conv1 = nn.Conv3d(
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/model.py:406]: It seems surf_to_vol_conv1 and surf_to_vol_conv2 are also not used anywhere. We can remove those.

File: physicsnemo/models/domino/model.py [hunk 5]
```diff
@@ -1316,14 +436,31 @@ def __init__(
...
+                        ),
                     )
                 )
 
+            self.solution_calculator_surf = SolutionCalculatorSurface(
+                num_variables=self.num_variables_surf,
+                num_sample_points=self.num_sample_points_surface,
+                use_surface_normals=self.use_surface_normals,
+                use_surface_area=self.use_surface_area,
+                noise_intensity=50,
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/model.py]: noise_intensity is not required for surface calculations. we can remove it.

File: physicsnemo/models/domino/solutions.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+        self.aggregation_model = aggregation_model
+        self.nn_basis = nn_basis
+
+        if self.encode_parameters:
+            if self.parameter_model is None:
+                raise ValueError(
+                    "Parameter model is required when encode_parameters is True"
+                )
+
+    def apply_parameter_encoding(
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/solutions.py]: Both surface and volume calculator use the same apply_parameter_encoding function. Can it be defined outside?

File: physicsnemo/models/domino/mlps.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+The DoMINO class contains an architecture to model both surface and
+volume quantities together as well as separately (controlled using
+the config.yaml file)
+"""
+
+import torch
+import torch.nn as nn
+
+
+class MLP(nn.Module):
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/mlps.py]: physicsnemo/models/layers has an MLP module. Can that be updated and used here? We could also define the FourierMLP (EncodingMLP in encodings.py) layer in physicsnemo/models/layers.

File: physicsnemo/models/domino/encodings.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+import torch
+import torch.nn as nn
+from einops import rearrange
+
+from .ball_query import BQWarp
+from .mlps import MLP, LocalPointConv
+
+
+def fourier_encode_vectorized(
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/encodings.py]: Can this function be moved to physicsnemo/models/layers so that it can be leveraged by other models.

File: physicsnemo/models/domino/ball_query.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+"""
+
+import torch
+import torch.nn as nn
+from einops import rearrange
+
+from physicsnemo.utils.neighbors import radius_search
+
+
+class BQWarp(nn.Module):
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/ball_query.py]: Can this be moved to physicsnemo/models/layers so that it can be leveraged by other models?

File: physicsnemo/models/domino/geometry_rep.py [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+        base_neurons = model_parameters.base_neurons
+        self.fourier_features = model_parameters.fourier_features
+        self.num_modes = model_parameters.num_modes
+
+        if self.fourier_features:
+            input_features_calculated = input_features * (1 + 2 * self.num_modes)
+        else:
+            input_features_calculated = input_features
+
+        self.fc1 = nn.Linear(input_features_calculated, base_neurons)
```
Comment (RishikeshRanade) [inline physicsnemo/models/domino/geometry_rep.py]: Can these layers be replaced with MLP or FourierMLP (EncodingMLP in encodings.py)?

**Review by RishikeshRanade — APPROVED** [review]
Great job on this. Thanks for working on it. I have left a few comments but overall looks good.

File: physicsnemo/models/domino/geometry_rep.py (hunk 1 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/geometry_rep.py]: Instead, I think it's best to use `get_activation` from physicsnemo.models.layers, it's more comprehensive.

File: physicsnemo/models/domino/encodings.py (hunk 2 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/encodings.py]: I added FourierMLP to physicsnemo.models.layers.fourier_layers.

File: physicsnemo/models/domino/model.py (hunk 3 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/model.py:555]: Yes, I can remove it.  Removing will break the numerical tests with the checkpoints, since creating the layer changes downstream random state.  I'll remove all the unused pieces last.

File: physicsnemo/models/domino/model.py (hunk 4 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/model.py:406]: Same as above - will remove last.

File: physicsnemo/models/domino/model.py (hunk 5 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/model.py]: Done.

File: physicsnemo/models/domino/mlps.py (hunk 7 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/mlps.py]: Done and done.

File: physicsnemo/models/domino/encodings.py (hunk 8 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/encodings.py]: Yes, named as just `fourier_encode`.

File: physicsnemo/models/domino/ball_query.py (hunk 9 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/ball_query.py]: Done.

File: physicsnemo/models/domino/geometry_rep.py (hunk 10 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/geometry_rep.py]: Done.

File: physicsnemo/models/domino/solutions.py (hunk 6 above)
Comment (coreyjadams) [reply physicsnemo/models/domino/solutions.py]: Done.

File: examples/cfd/external_aerodynamics/domino/src/conf/config.yaml [hunk 11]
```diff
@@ -65,7 +65,7 @@ variables:
       reference: [38.89] # vector [30, 0, 0] should be specified as [30], while [30, 30, 0] should be [30, 30].
     air_density:
       type: scalar
-      reference: 1.226
+      reference: 1.0
```
Comment (Alexey-Kamenev) [inline examples/cfd/external_aerodynamics/domino/src/conf/config.yaml:68]: So the reference is not a sea level density anymore?

File: examples/cfd/external_aerodynamics/domino/src/test.py [hunk 12]
```diff
@@ -554,22 +563,11 @@ def main(cfg: DictConfig):
...
-                )
-            else:
-                pos_surface_center_of_mass = surface_coordinates - center_of_mass
+            
+            pos_surface_center_of_mass = surface_coordinates - center_of_mass_normalized
 
-            surface_coordinates = normalize(surface_coordinates, s_max, s_min)
-            surface_neighbors = normalize(surface_neighbors, s_max, s_min)
-            surf_grid = normalize(surf_grid, s_max, s_min)
+            # surface_coordinates = normalize(surface_coordinates, s_max, s_min)
```
Comment (Alexey-Kamenev) [inline examples/cfd/external_aerodynamics/domino/src/test.py:569]: Remove commented out code or add clarifying comment?

File: physicsnemo/models/layers/mlp_layers.py [hunk 13]
```diff
@@ -17,28 +17,83 @@
 import torch
 from torch import nn
 
+from .activations import get_activation
+
+# class Mlp(nn.Module):
```
Comment (Alexey-Kamenev) [inline physicsnemo/models/layers/mlp_layers.py:22]: Remove?

**Review by Alexey-Kamenev — APPROVED** [review]
LGTM!

**Comment by coreyjadams** [comment]
FYI, this has been merged into the `domino` branch to do some testing w/ the performance optimizations.  Thanks for the reviews!
